hpx/parallel/datapar/detail/iterator_helpers.hpp

```diff
@@ -0,0 +1,312 @@
+                iter_(iter)
+            {
+            }
+            ~store_on_exit()
+            {
+                value_.store(std::addressof(*iter_), Vc::Aligned);
+            }
+
+            V* operator&() { return &value_; }
+            V const* operator&() const { return &value_; }
+
+            V value_;
+            Iter iter_;
+        };
+
+        template <typename Iter, typename V>
+        struct store_on_exit<Iter, V,
+            typename std::enable_if<
+                std::is_const<
+                    typename std::iterator_traits<Iter>::value_type
+                >::value
+            >::type>
+        {
+            store_on_exit(Iter const& iter)
+              : value_(std::addressof(*iter), Vc::Aligned)
+            {
+            }
+
+            V* operator&() { return &value_; }
+            V const* operator&() const { return &value_; }
+
+            V value_;
+        };
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename Iter, typename Enable = void>
+        struct datapar_loop_step
+        {
+            typedef typename std::iterator_traits<Iter>::value_type value_type;
+
+            typedef Vc::Scalar::Vector<value_type> V1;
+            typedef Vc::Vector<value_type> V;
+
+            template <typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            static typename std::result_of<F&&(V1*)>::type
+            call1(F && f, Iter& it)
+            {
+                store_on_exit<Iter, V1> tmp(it);
+                it += V1::Size;
+                return hpx::util::invoke(f, &tmp);
+            }
+
+            template <typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            static typename std::result_of<F&&(V*)>::type
+            callv(F && f, Iter& it)
+            {
+                store_on_exit<Iter, V> tmp(it);
+                it += V::Size;
+                return hpx::util::invoke(f, &tmp);
+            }
+        };
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename Iter1, typename Iter2>
+        struct datapar_loop_step2
+        {
+            typedef typename std::iterator_traits<Iter1>::value_type value1_type;
+            typedef typename std::iterator_traits<Iter2>::value_type value2_type;
+
+            typedef Vc::Scalar::Vector<value1_type> V11;
+            typedef Vc::Scalar::Vector<value2_type> V12;
+
+            typedef Vc::Vector<value1_type> V1;
+            typedef Vc::Vector<value2_type> V2;
+
+            template <typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            static typename std::result_of<F&&(V11*, V12*)>::type
+            call1(F && f, Iter1& it1, Iter2& it2)
+            {
+                V11 tmp1(std::addressof(*it1), Vc::Aligned);
+                V12 tmp2(std::addressof(*it2), Vc::Aligned);
+                std::advance(it1, V11::Size);
+                std::advance(it2, V12::Size);
```

Comment:
The sizes are checked [here](https://github.com/STEllAR-GROUP/hpx/blob/datapar/hpx/parallel/datapar/detail/iterator_helpers.hpp#L50). If the sizes are different we fall back to sequential execution anyways.

> Also, since it is scalar, shouldn't V11::Size == V12::Size == 1?

Yes, you're right. a simple `++it1, ++it2` should do the trick here.